File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        // iterate throguht he list, store the the head as temp_prev
        // go to next node, store node's next as temp_next. set node->next=temp_prev
        // repeat
        ListNode* temp_prev = nullptr;
        ListNode* curr = head;
        ListNode* temp_next = curr->next;
        while (curr != nullptr){
            temp_next = curr->next;
            curr->next = temp_prev;
            temp_prev = curr;
            curr = temp_next;
        }
        return temp_prev;
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        left = dummy
        right = dummy

        for i in range(n):
            right = right.next

        while right.next:
            left = left.next
            right = right.next

        left.next = left.next.next

        return dummy.next
        # two pointers method
        # left ummy, right at head, with a loop move right n steps forward
        # after that, the gap between left and right is n + 1,
        # when right reaches the end, left is pointing to the nth node 
        # left.next = left.next.next removes the nth node
        # return dummy
        